tests/unit/fauxdbi.py

```diff
@@ -54,45 +272,150 @@ def executemany(self, operation, parameters_list):
     def close(self):
         self.cursor.close()
 
-    def fetchone(self):
-        return self.cursor.fetchone()
+    def _fix_pickled(self, row):
+        if row is None:
+            return row
 
-    def fetchmany(self, size=None):
-        self.cursor.fetchmany(size or self.arraysize)
+        return [
+            (
+                pickle.loads(v.encode("latin1"))
+                if isinstance(v, str) and v[:2] == "\x80\x04" and v[-1] == "."
+                else pickle.loads(base64.b16decode(v))
+                if isinstance(v, str) and v[:4] == "8004" and v[-2:] == "2E"
+                else v
+            )
+            for d, v in zip(self.description, row)
+        ]
+
+    def fetchone(self):
+        return self._fix_pickled(self.cursor.fetchone())
 
     def fetchall(self):
-        return self.cursor.fetchall()
+        return map(self._fix_pickled, self.cursor)
 
-    def setinputsizes(self, sizes):
-        pass
 
-    def setoutputsize(self, size, column=None):
-        pass
+class attrdict(dict):
+    def __setattr__(self, name, val):
+        self[name] = val
+
+    def __getattr__(self, name):
+        if name not in self:
+            self[name] = attrdict()
+        return self[name]
 
 
 class FauxClient:
-    def __init__(self, client, connection):
-        self._client = client
-        self.project = client.project
-        self.connection = connection
+    def __init__(self, project_id=None, default_query_job_config=None, *args, **kw):
+
+        if project_id is None:
+            if default_query_job_config is not None:
+                project_id = default_query_job_config.default_dataset.project
+            else:
+                project_id = "authproj"  # we would still have gotten it from auth.
+
+        self.project = project_id
+        self.tables = attrdict()
+
+    @staticmethod
+    def _row_dict(row, cursor):
+        result = {d[0]: value for d, value in zip(cursor.description, row)}
+        return result
+
+    def _get_field(
+        self,
+        type,
+        name=None,
+        notnull=None,
+        mode=None,
+        description=None,
+        fields=(),
+        columns=None,
+        **_,
+    ):
+        if columns:
+            custom = columns.get(name)
+            if custom:
+                return self._get_field(
+                    **dict(name=name, type=type, notnull=notnull, **custom)
```

Comment:
We express the kwargs this way to allow custom values to overrise
name, type and nonnull, if necessary.

I added some comments.

Reply:
Oops. I had in my head: `dict(dict(name='x'), **dict(name='y'))`

Since we're not relying on overriding (I probably was or planned to at some point), I just got rid of it.

Thanks!